src/runtime_src/core/tools/xbutil2/SubCmdValidate.cpp

```diff
@@ -284,91 +275,149 @@ runTestCase( const std::shared_ptr<xrt_core::device>& _dev, const std::string& p
-      logger(_ptTest, "Error", "Please check if the platform package is installed correctly");
-      _ptTest.put("status", test_token_failed);
-      return;
-    }
+    std::string test_name = file;
+    if (test_map.find(file) != test_map.end())
+      test_name = test_map.find(file)->second;
 
-    // log testcase path for debugging purposes
-    logger(_ptTest, "Testcase", xrtTestCasePath);
+    xrtTestCasePath = "/opt/xilinx/xrt/test/" + test_name;
+  } else {
+    xrtTestCasePath = "/opt/xilinx/xrt/test/" + file;
+  }
+  // Parse if the file exists here
+  boost::filesystem::path xrt_path(xrtTestCasePath);
+  if (!boost::filesystem::exists(xrt_path)) {
+    logger(_ptTest, "Error", boost::str(boost::format("Failed to find %s") % xrtTestCasePath));
+    logger(_ptTest, "Error", "Please check if the platform package is installed correctly");
+    _ptTest.put("status", test_token_failed);
+    throw xrt_core::error(std::errc::operation_canceled);
+  }
+  return xrtTestCasePath;
+}
 
-    std::vector<std::string> args = { "-p", test_dir,
-                                      "-d", xrt_core::query::pcie_bdf::to_string(xrt_core::device_query<xrt_core::query::pcie_bdf>(_dev)) };
-    try {
-      int exit_code = XBU::runScript("sh", xrtTestCasePath, args, "Running Test", MAX_TEST_DURATION, os_stdout, os_stderr);
-      if (exit_code == EOPNOTSUPP) {
-        _ptTest.put("status", test_token_skipped);
-      }
-      else if (exit_code == EXIT_SUCCESS) {
+/*
+ * Spawn a testcase process.
+ */
+bool
+spawn_testcase(boost::property_tree::ptree& _ptTest, const std::string& xrtTestCasePath,
+               std::ostringstream & os_stdout, std::ostringstream & os_stderr,
+               std::vector<std::string> args, const test_mode& mode)
+{
+  int exit_code;
+  try {
+    switch (mode) {
+      case test_mode::json:
+        exit_code = XBU::runScript("sh", xrtTestCasePath, args, "Running Test", MAX_TEST_DURATION, os_stdout, os_stderr);
+        break;
+      case test_mode::support:
+        exit_code = XBU::runScript("sh", xrtTestCasePath, args, "Checking Test Support", MAX_TEST_DURATION, os_stdout, os_stderr);
+        break;
+      case test_mode::no_json:
+        if (_ptTest.get<std::string>("file").find(".exe") != std::string::npos)
+          exit_code = XBU::runScript("", xrtTestCasePath, args, "Running Test", MAX_TEST_DURATION, os_stdout, os_stderr);
+        else
+          exit_code = XBU::runScript("python", xrtTestCasePath, args, "Running Test", MAX_TEST_DURATION, os_stdout, os_stderr);
+        break;
+      default:
+        throw std::runtime_error("Invalid test mode specified"); //Mode was not specified correctly.
+    }
+
+    if (exit_code == EOPNOTSUPP) {
+      logger(_ptTest, "Details", "Test is not supported on current device.");
+      _ptTest.put("status", test_token_skipped);
+      return false;
+    } else if (exit_code == EXIT_SUCCESS) {
```

Comment:
Why the else statement since line 327 exited the method.
```suggestion
    } 
    
    if (exit_code == EXIT_SUCCESS) {
```